dll/shellext/zipfldr/CZipExtract.cpp

```diff
@@ -103,20 +150,74 @@ class CZipExtract :
             if (m_pExtract->m_DirectoryChanged)
                 UpdateDirectory();
 
-            if (!m_pExtract->Extract(m_hWnd, GetDlgItem(IDC_PROGRESS)))
+            if (m_hExtractionThread != INVALID_HANDLE_VALUE)
             {
-                /* Extraction failed, do not go to the next page */
+                CloseHandle(m_hExtractionThread);
+                m_hExtractionThread = INVALID_HANDLE_VALUE;
+            }
+            m_hExtractionThread = CreateThread(
+                NULL, 0,
+                &CExtractSettingsPage::ExtractEntry,
+                this,
+                CREATE_SUSPENDED,
+                NULL);
+
+            if (!m_hExtractionThread)
+            {
+                /* Extraction thread creation failed, do not go to the next page */
                 SetWindowLongPtr(DWLP_MSGRESULT, -1);
 
                 ::EnableWindow(GetDlgItem(IDC_BROWSE), TRUE);
                 ::EnableWindow(GetDlgItem(IDC_DIRECTORY), TRUE);
                 ::EnableWindow(GetDlgItem(IDC_PASSWORD), TRUE);
                 SetWizardButtons(PSWIZB_NEXT);
 
+                m_hExtractionThread = INVALID_HANDLE_VALUE;
+
                 return TRUE;
             }
+            ResumeThread(m_hExtractionThread);
+            return -1;
```

Comment:
The documentation is tricksy here. "Returning" `-1` actually means `SetWindowLong(DWL_MSGRESULT, -1); return TRUE;`
See the Remarks section in https://docs.microsoft.com/en-us/windows/win32/controls/psn-wiznext